From 70e3d626378e12ea50d76dfda50311c8bb4a2a78 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?J=C3=B6rg=20Krause?= <joerg.krause@embedded.rocks>
Date: Tue, 5 Dec 2017 11:29:44 +0100
Subject: [PATCH] Fix segmentation fault in http_MakeMessage

When upmpdcli is linked with libupnp 1.6.24 it fails with an
segmentation fault, but linking with libupnp worked just fine.

git bisect shows that commit 9c2e8ec8a0291ebe81959009e2f78edbdb47ced5 is
the bad one as the variable `extras` is not properly checked before
attempting to used it as a pointer.

Asure `extras` is not Null before using it.
---
 upnp/src/genlib/net/http/httpreadwrite.c | 16 +++++++++-------
 1 file changed, 9 insertions(+), 7 deletions(-)

diff --git a/upnp/src/genlib/net/http/httpreadwrite.c b/upnp/src/genlib/net/http/httpreadwrite.c
index 68cd4baa..57ab8051 100644
--- a/upnp/src/genlib/net/http/httpreadwrite.c
+++ b/upnp/src/genlib/net/http/httpreadwrite.c
@@ -1662,14 +1662,16 @@ int http_MakeMessage(membuffer *buf, int http_major_version,
 			struct Extra_Headers *extras;
 			/* array of extra headers */
 			extras = (struct Extra_Headers *) va_arg(argp, struct Extra_Headers *);
-			while (extras->name) {
-				if (extras->resp) {
-					if (membuffer_append(buf, extras->resp, strlen(extras->resp)))
-						goto error_handler;
-					if (membuffer_append(buf, "\r\n", (size_t)2))
-						goto error_handler;
+			if (extras) {
+				while (extras->name) {
+					if (extras->resp) {
+						if (membuffer_append(buf, extras->resp, strlen(extras->resp)))
+							goto error_handler;
+						if (membuffer_append(buf, "\r\n", (size_t)2))
+							goto error_handler;
+					}
+					extras++;
 				}
-				extras++;
 			}
 		}
 		if (c == 's') {
